src/mime.cr

```diff
@@ -0,0 +1,200 @@
+    from_extension(extension) { nil }
+  end
+
+  # Looks up the MIME type associated with *extension*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Runs the fiven block if *extension* is not registered.
+  def self.from_extension(extension : String, &block)
+    initialize_types
+
+    @@types.fetch(extension) { @@types_lower.fetch(extension.downcase) { yield extension } }
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Returns *default* if *extension* is not registered.
+  def self.from_filename(filename : String, default) : String
+    from_extension(File.extname(filename), default)
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Raises `KeyError` if extension is not registered.
+  def self.from_filename(filename : String) : String
+    from_extension(File.extname(filename))
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Returns `nil` if extension is not registered.
+  def self.from_filename?(filename : String) : String?
+    from_extension?(File.extname(filename))
+  end
+
+  # Looks up the MIME type associated with the extension in *filename*.
+  #
+  # A case sensitive search is tried first, if this yields no result, it is
+  # matched case-insensitive. Runs the fiven block if extension is not registered.
+  def self.from_filename(filename : String, &block)
+    from_extension(File.extname(filename)) { |extension| yield extension }
+  end
+
+  # Register *type* for *extension*.
+  #
+  # *extension* must start with a dot (`.`) and must not contain any null bytes.
+  def self.register(extension : String, type : String) : Nil
+    raise ArgumentError.new("Extension does not start with a dot: #{extension.inspect}") unless extension.starts_with?('.')
+    extension = extension.check_no_null_byte
+
+    initialize_types
+
+    mediatype = parse_media_type(type) || raise Error.new "Invalid media type: #{type}"
+
+    @@types[extension] = type
+    @@types_lower[extension.downcase] = type
+
+    if mediatype
+      array = @@extensions[mediatype] ||= [] of String
+      array << extension
+    end
+
+    nil
+  end
+
+  # Returns all extensions registered for *type*.
+  def self.extensions(type : String) : Array(String)
+    @@extensions.fetch(type) { [] of String }
+  end
+
+  # tspecial as defined by RFC 1521 and RFC 2045
+  private TSPECIAL_CHARACTERS = {'(', ')', '<', '>', '@', ',', ';', ':', '\\', '"', '/', '[', ']', '?', '='}
+
+  # :nodoc:
+  def self.parse_media_type(type : String) : String?
```

Comment:
Must be private. It's only public for testing purposes, but internal implementation detail musn't be tested directly. Test `MIME.register` with various cases instead.

Reply:
Easier != what should be done.

I don't deny the tests were helpful to design the feature, but either `#parse_media_type` as a value by itself, and thus should be public & tested, or it's an implementation detail and associated tests can be dropped.